src/WorkItems/Logging/ApplicationInsightsTelemetryInitializer.cs

```diff
@@ -0,0 +1,35 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using Microsoft.ApplicationInsights.Channel;
+using Microsoft.ApplicationInsights.Extensibility;
+
+namespace Microsoft.WorkItems.Logging
+{
+    /// <summary>
+    /// Assigns a GUID as the operation_Id for the Application Insights traces.
+    /// </summary>
+    internal class ApplicationInsightsTelemetryInitializer : ITelemetryInitializer
+    {
+        private readonly string m_operationId;
+
+        public ApplicationInsightsTelemetryInitializer()
+        {
+            m_operationId = Guid.NewGuid().ToString();
+        }
+
+        public void Initialize(ITelemetry telemetry)
+        {
+            if (telemetry?.Context?.Operation == null)
+            {
+                throw new ArgumentNullException(nameof(telemetry));
+            }
+
+            if (string.IsNullOrEmpty(telemetry.Context.Operation.Id))
```

Comment:
>if (string.IsNullOrEmpty(telemetry.Context.Operation.Id)) [](start = 12, length = 57)

Why is this right? i.e., that we preserve the id of the incoming thing when we call initialize. Why don't we, for example, reset m_operationId in this case? i.e., how is it legal/useful for the ids to diverge.

Reply:
I think the risk of mixing the operation ids is that there's something else logging telemetry (perhaps capturing perf counters) and they are setting their own operation id (there's a chain of initializers, so they could insert themselves in the chain.) In this case, the operation ids are supposed to be different.

---
In reply to: [396832352](https://github.com/microsoft/sarif-sdk/pull/1820#discussion_r396832352) [](ancestors = 396832352)